runtime/shared/shrclssup.c

```diff
@@ -330,6 +330,14 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
 				if (nonfatal) {
 					return J9VMDLLMAIN_OK;
 				} else {
+#if defined(J9ZOS390)
+					if ((J9VMDLLMAIN_SILENT_EXIT_VM == rc)
+						&& J9_ARE_ALL_BITS_SET(vm->sharedCacheAPI->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_PERSISTENT_CACHE)
+					) {
+						/* Call j9shr_guaranteed_exit() to ensure any modification to the cache is flushed to the file on disk. */
+						j9shr_guaranteed_exit(vm, FALSE);
```

Comment:
I don't think we want to do this unconditionally, for all utilities. It writes to the cache, which is undesirable for utilities such as printStats. Better to do this only for utilities that made a modification.

Reply:
I went through all the CML utilities options, `printStats` (and its variants), adjusting softmx/aot space/jit space limits and `revalidateAotMethods/invalidateAotMethods=`  are the ones that we need to consider for this issue.

For `printStats` and its variants, we already set the readonly runtime flag, so there won't be a modification to the cache.

https://github.com/eclipse-openj9/openj9/blob/5dd147e65442c8c954b6b72260b57c83a6b25717/runtime/shared_common/shrinit.cpp#L3297-L3308

But I can still add a check of `J9SHR_RUNTIMEFLAG_ENABLE_READONLY` here if you want.

Reply:
Sure, pls add that to make it explicit.

Reply:
A comment that printStats sets this flag would also be nice.